beluga/include/beluga/sensor/ndt_sensor_model.hpp

```diff
@@ -0,0 +1,234 @@
+  for (const Eigen::Vector2d& col : points) {
+    cell_grid[(col / resolution).cast<int>()].emplace_back(col);
+  }
+
+  for (const auto& [cell, points_in_cell] : cell_grid) {
+    if (points_in_cell.size() < kMinPointsPerCell) {
+      continue;
+    }
+    ret.push_back(fit_points(points_in_cell));
+  }
+
+  return ret;
+}
+
+/// NDT sensor model for range finders.
+/**
+ * This class satisfies \ref SensorModelPage.
+ *
+ * \tparam SparseGridT Type representing a sparse NDT grid, as a specialization of 'beluga::SparseValueGrid'.
+ */
+template <typename SparseGridT>
+class NDTSensorModel {
+ public:
+  static_assert(std::is_same_v<SparseValueGrid<typename SparseGridT::map_type>, SparseGridT>);
+  static_assert(std::is_same_v<typename SparseGridT::mapped_type, NDTCell>);
+  /// State type of a particle.
+  using state_type = Sophus::SE2d;
+  /// Weight type of the particle.
+  using weight_type = double;
+  /// Measurement type of the sensor: a point cloud for the range finder.
+  using measurement_type = std::vector<Eigen::Vector2d>;
+  /// Map representation type.
+  using map_type = SparseGridT;
+  /// Parameter type that the constructor uses to configure the NDT sensor model.
+  using param_type = NDTModelParam;
+
+  /// Constructs a NDTSensorModel instance.
+  /**
+   * \param params Parameters to configure this instance.
+   *  See beluga::NDTModelParam for details.
+   * \param cells_data Sparse grid representing an NDT map, used for calculating importance weights for the different
+   * particles.
+   */
+  NDTSensorModel(param_type param, SparseGridT cells_data)
+      : params_{std::move(param)}, cells_data_{std::move(cells_data)} {}
+
+  /// Returns a state weighting function conditioned on 2D lidar hits.
+  /**
+   * \param points 2D lidar hit points in the reference frame of particle states.
+   * \return a state weighting function satisfying \ref StateWeightingFunctionPage
+   *  and borrowing a reference to this sensor model (and thus their lifetime are bound).
+   */
+  [[nodiscard]] auto operator()(measurement_type&& points) const {
+    return [this, cells = to_cells(points, cells_data_.resolution())](const state_type& state) -> weight_type {
+      return std::transform_reduce(
+          cells.cbegin(), cells.cend(), 1.0, std::plus{},
+          [this, state](const NDTCell& ndt_cell) { return likelihood_at(state * ndt_cell); });
+    };
+  }
+
+  /// Returns the L2 likelihood scaled by 'd1' and 'd2' set in the parameters for this instance for 'measurement, or
+  /// 'params_.min_likelihood' if the cell corresponding to 'measurement' doesn't exist in the map.
+  [[nodiscard]] double likelihood_at(const NDTCell& measurement) const {
+    const auto maybe_cell = cells_data_.data_near(measurement.mean);
+    if (!maybe_cell.has_value()) {
+      return params_.minimum_likelihood;
+    }
+    return maybe_cell->likelihood_at(measurement, params_.d1, params_.d2);
+  }
```

Comment:
@serraramiro1 does this need to be a public method? Does it need to be a sensor model method?

Reply:
It makes testing way easier and thorough, in my opinion.

I'm down to make it private + friend test if needed but I'd rather keep some version of it.